ci/scripts/run-codebuild/libs/codebuild-runner.js

```diff
@@ -0,0 +1,158 @@
+// ourselves and go back to the past and do all the pagination
+import { CloudWatchLogsClient, GetLogEventsCommand } from '@aws-sdk/client-cloudwatch-logs'
+import { CodeBuildClient, StartBuildCommand, BatchGetBuildsCommand } from '@aws-sdk/client-codebuild'
+
+import { promisify } from 'util'
+
+const sleep = promisify(setTimeout)
+
+export class CodeBuildError extends Error {
+  constructor (message, buildInfo) {
+    super(message)
+    this.name = 'CodeBuildError'
+    this.buildInfo = buildInfo
+  }
+}
+
+export class CodeBuildRunner {
+  constructor (projectName, sourceVersion, secondarySourceVersions, environmentVariables, region = 'eu-west-1', defaultSleepTime = 5000) {
+    this.defaultSleepTime = defaultSleepTime
+
+    this.buildCommandInput = {
+      projectName: projectName,
+      sourceVersion: sourceVersion,
+      secondarySourcesVersionOverride: secondarySourceVersions,
+      environmentVariablesOverride: environmentVariables
+    }
+
+    this.codeBuildClient = new CodeBuildClient({ region: region })
+    this.cloudWatchLogsClient = new CloudWatchLogsClient({ region: region })
+
+    this.buildId = null
+  }
+
+  async runBuildAndLog () {
+    let build = await this._startBuild()
+
+    console.log(`Sent command to start build ${build.buildNumber} for ${build.projectName} at ${build.startTime}`)
+
+    await this._waitForBuildToProduceLogs()
+
+    console.log(`Build ${this.buildId} is running, logs follow`)
+    console.log('='.repeat(80))
+
+    await sleep(this.defaultSleepTime)
+
+    build = await this._logWhileWaitingForBuildToComplete()
+
+    await sleep(this.defaultSleepTime)
+
+    console.log('='.repeat(80))
+    console.log(`Build finished with status ${build.buildStatus}`)
+
+    return await this.getBuildInfo()
+  }
+
+  async getBuildInfo () {
+    const command = new BatchGetBuildsCommand({ ids: [this.buildId] })
+
+    const commandResponse = await this.codeBuildClient.send(command)
+
+    return commandResponse.builds[0]
+  }
+
+  async _startBuild () {
+    const command = new StartBuildCommand(this.buildCommandInput)
+
+    const startBuildResponse = await this.codeBuildClient.send(command)
+
+    const build = startBuildResponse.build
+
+    this.buildId = build.id
+
+    return build
+  }
+
+  async _waitForBuildToProduceLogs () {
+    console.log(`Waiting for build ${this.buildId} to be in progress`)
+
+    while (true) {
+      const build = await this.getBuildInfo()
+
+      if (build.buildStatus !== 'IN_PROGRESS') {
+        throw new CodeBuildError(`Codebuild finished with status ${build.buildStatus} before producing any logs!`, build)
+      }
+
+      if (typeof build.logs.groupName !== 'undefined') {
+        // Newline gives us a blank line after all the progress .'s
+        console.log(`\nLogs are now availabile at ${build.logs.deepLink} in group ${build.logs.groupName} and stream ${build.logs.streamName}`)
```

Comment:
Typo - `availabile`

Reply:
Ah yeah. That's weird, I don't know why but my spellcheck sees this as a valid word O_o It's clearly not! I wonder if I accidentally added it to a custom wordlist 🤔

Reply:
Sorted 👍